三种哈希算法比较图像的相似性

In [ ]:
# -*- coding: utf-8 -*-
import pandas as pd
import cv2
import time
import numpy as np

In [13]:

def pHash(img,leng=32,wid=32):
    img = cv2.resize(img, (leng, wid))   
    gray = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)
    dct = cv2.dct(np.float32(gray))
    dct_roi = dct[0:8, 0:8]            
    avreage = np.mean(dct_roi)
    phash_01 = (dct_roi>avreage)+0
    phash_list = phash_01.reshape(1,-1)[0].tolist()
    hash = ''.join([str(x) for x in phash_list])
    return hash

def dHash(img,leng=9,wid=8):
    img=cv2.resize(img,(leng, wid))
    image=cv2.cvtColor(img,cv2.COLOR_BGR2GRAY)
    #每行前一个像素大于后一个像素为1，相反为0，生成哈希
    hash=[]
    for i in range(wid):
        for j in range(wid):
            if image[i,j]>image[i,j+1]:
                hash.append(1)
            else:
                hash.append(0)
    return hash

def aHash(img,leng=8,wid=8):
    img=cv2.resize(img,(leng, wid))
    image=cv2.cvtColor(img,cv2.COLOR_BGR2GRAY)
    avreage = np.mean(image)                           
    hash = [] 
    for i in range(image.shape[0]): 
        for j in range(image.shape[1]): 
            if image[i,j] >= avreage: 
                hash.append(1) 
            else: 
                hash.append(0) 
    return hash

def Hamming_distance(hash1,hash2):
    num = 0
    for index in range(len(hash1)):
        if hash1[index] != hash2[index]:
            num += 1
    return num 

if __name__ == '__main__':
    for i in range(0,50):
        image1 = f'F:/cifar-10-batches-py/pic/test/{i}.jpg'
        #image2 = f'F:/poison_cifar10/test_pic/{i}.png'
        image2 = f'F:/patched-cifar-10/test_pic/{i}.png'
        image1 = cv2.imread(image1)
        image2 = cv2.imread(image2)
        
        d_dist = Hamming_distance(dHash(image1),dHash(image2))

        p_dist = Hamming_distance(pHash(image1),pHash(image2))

        a_dist = Hamming_distance(aHash(image1),aHash(image2))
        print(d_dist)
    for i in range(0,50):
        image1 = f'F:/cifar-10-batches-py/pic/test/{i}.jpg'
        image2 = f'F:/poison_cifar10/test_pic/{i}.png'
        #image2 = f'F:/patched-cifar-10/test_pic/{i}.png'
        image1 = cv2.imread(image1)
        image2 = cv2.imread(image2)
        
        d_dist = Hamming_distance(dHash(image1),dHash(image2))

        p_dist = Hamming_distance(pHash(image1),pHash(image2))

        a_dist = Hamming_distance(aHash(image1),aHash(image2))
        print(1 - d_dist * 1.0 / 64)
        #print('a_dist is '+'%d' % a_dist + ', similarity is ' +'%f' % (1 - a_dist * 1.0 / 64))
        #print('p_dist is '+'%d' % p_dist + ', similarity is ' +'%f' % (1 - p_dist * 1.0 / 64))
        #print('d_dist is '+'%d' % d_dist + ', similarity is ' +'%f' % (1 - d_dist * 1.0 / 64))


8
6
9
7
7
9
7
6
4
6
11
5
6
2
10
7
5
4
7
7
6
15
8
9
5
10
1
5
6
7
11
7
10
4
7
5
11
7
15
9
8
6
10
7
19
6
5
8
9
6
0.875
0.828125
0.8125
0.796875
0.921875
0.796875
0.890625
0.84375
0.859375
0.890625
0.703125
0.90625
0.859375
0.96875
0.8125
0.796875
0.921875
0.921875
0.796875
0.828125
0.875
0.5625
0.859375
0.8125
0.859375
0.859375
0.890625
0.875
0.953125
0.84375
0.765625
0.796875
0.90625
0.953125
0.765625
0.890625
0.765625
0.84375
0.765625
0.90625
0.78125
0.796875
0.828125
0.859375
0.59375
0.71875
0.875
0.84375
0.765625
0.9375


cifar10 - benign - poison - aHash
aHash distance
17
3
3
1
4
5
7
7
7
7
3
14
2
4
10
6
2
9
4
5
2
0
8
6
3
6
3
2
6
7
5
3
6
6
2
1
5
6
9
5
3
7
4
8
8
7
2
10
7
2
aHash similarity
0.734375
0.953125
0.953125
0.984375
0.9375
0.921875
0.890625
0.890625
0.890625
0.890625
0.953125
0.78125
0.96875
0.9375
0.84375
0.90625
0.96875
0.859375
0.9375
0.921875
0.96875
1.0
0.875
0.90625
0.953125
0.90625
0.953125
0.96875
0.90625
0.890625
0.921875
0.953125
0.90625
0.90625
0.96875
0.984375
0.921875
0.90625
0.859375
0.921875
0.953125
0.890625
0.9375
0.875
0.875
0.890625
0.96875
0.84375
0.890625
0.96875

cifar10 - benign - poison - dHash
8
11
12
13
5
13
7
10
9
7
19
6
9
2
12
13
5
5
13
11
8
28
9
12
9
9
7
8
3
10
15
13
6
3
15
7
15
10
15
6
14
13
11
9
26
18
8
10
15
4
0.875
0.828125
0.8125
0.796875
0.921875
0.796875
0.890625
0.84375
0.859375
0.890625
0.703125
0.90625
0.859375
0.96875
0.8125
0.796875
0.921875
0.921875
0.796875
0.828125
0.875
0.5625
0.859375
0.8125
0.859375
0.859375
0.890625
0.875
0.953125
0.84375
0.765625
0.796875
0.90625
0.953125
0.765625
0.890625
0.765625
0.84375
0.765625
0.90625
0.78125
0.796875
0.828125
0.859375
0.59375
0.71875
0.875
0.84375
0.765625
0.9375

cifar10 - benign - poison - pHash
4
3
5
1
1
2
5
1
2
1
3
1
4
1
2
1
3
4
1
1
1
1
3
1
1
0
0
0
3
2
2
2
0
1
4
1
1
0
2
5
1
4
3
1
7
1
2
1
2
3
0.9375
0.953125
0.921875
0.984375
0.984375
0.96875
0.921875
0.984375
0.96875
0.984375
0.953125
0.984375
0.9375
0.984375
0.96875
0.984375
0.953125
0.9375
0.984375
0.984375
0.984375
0.984375
0.953125
0.984375
0.984375
1.0
1.0
1.0
0.953125
0.96875
0.96875
0.96875
1.0
0.984375
0.9375
0.984375
0.984375
1.0
0.96875
0.921875
0.984375
0.9375
0.953125
0.984375
0.890625
0.984375
0.96875
0.984375
0.96875
0.953125

cifar10 - benign - patch - aHash
18
4
4
4
6
6
5
6
6
8
1
15
2
4
11
8
2
7
5
6
2
1
6
7
3
5
3
5
5
5
5
5
3
6
2
1
5
6
5
6
3
7
5
7
9
7
2
7
10
5
0.734375
0.953125
0.953125
0.984375
0.9375
0.921875
0.890625
0.890625
0.890625
0.890625
0.953125
0.78125
0.96875
0.9375
0.84375
0.90625
0.96875
0.859375
0.9375
0.921875
0.96875
1.0
0.875
0.90625
0.953125
0.90625
0.953125
0.96875
0.90625
0.890625
0.921875
0.953125
0.90625
0.90625
0.96875
0.984375
0.921875
0.90625
0.859375
0.921875
0.953125
0.890625
0.9375
0.875
0.875
0.890625
0.96875
0.84375
0.890625
0.96875

cifar10 - benign - patch - dHash
8
6
9
7
7
9
7
6
4
6
11
5
6
2
10
7
5
4
7
7
6
15
8
9
5
10
1
5
6
7
11
7
10
4
7
5
11
7
15
9
8
6
10
7
19
6
5
8
9
6
0.875
0.828125
0.8125
0.796875
0.921875
0.796875
0.890625
0.84375
0.859375
0.890625
0.703125
0.90625
0.859375
0.96875
0.8125
0.796875
0.921875
0.921875
0.796875
0.828125
0.875
0.5625
0.859375
0.8125
0.859375
0.859375
0.890625
0.875
0.953125
0.84375
0.765625
0.796875
0.90625
0.953125
0.765625
0.890625
0.765625
0.84375
0.765625
0.90625
0.78125
0.796875
0.828125
0.859375
0.59375
0.71875
0.875
0.84375
0.765625
0.9375

cifar10 - benign - patch - pHash
10
3
5
2
6
3
4
7
9
2
3
1
6
5
3
3
4
2
1
8
7
1
8
4
2
0
2
1
2
7
4
2
5
2
5
6
1
2
1
8
3
14
1
4
6
0
3
1
6
6
0.9375
0.953125
0.921875
0.984375
0.984375
0.96875
0.921875
0.984375
0.96875
0.984375
0.953125
0.984375
0.9375
0.984375
0.96875
0.984375
0.953125
0.9375
0.984375
0.984375
0.984375
0.984375
0.953125
0.984375
0.984375
1.0
1.0
1.0
0.953125
0.96875
0.96875
0.96875
1.0
0.984375
0.9375
0.984375
0.984375
1.0
0.96875
0.921875
0.984375
0.9375
0.953125
0.984375
0.890625
0.984375
0.96875
0.984375
0.96875
0.953125

cifar10 - benign - DBA - aHash

cifar10 - benign - DBA - dHash

cifar10 - benign - DBA - pHash